File: backend/app/schemas/agent.py
"""
Agent schemas for API request/response validation.
"""

from datetime import datetime
from typing import Optional

from pydantic import BaseModel, Field


class InteractionRules(BaseModel):
    """Agent interaction rules configuration."""

    can_challenge: bool = Field(default=True, description="Whether agent can challenge others")
    can_be_challenged: bool = Field(default=True, description="Whether agent accepts challenges")
    defer_to: list[str] = Field(default_factory=list, description="Agent IDs to defer to in their domain")


class AgentBase(BaseModel):
    """Base agent schema with common fields."""

    name: str = Field(..., min_length=1, max_length=100, description="Agent name")
    avatar: Optional[str] = Field(default=None, max_length=500, description="Avatar URL")
    description: Optional[str] = Field(default=None, description="Agent description")
    tags: list[str] = Field(default_factory=list, description="Searchable tags")

    # Core configuration
    system_prompt: str = Field(..., min_length=1, description="System prompt for the agent")
    model_id: Optional[str] = Field(default=None, description="Specific model to use")
    temperature: float = Field(default=0.7, ge=0, le=2, description="LLM temperature")
    max_tokens: int = Field(default=2048, ge=1, le=128000, description="Max output tokens")

    # Capabilities
    tools: list[str] = Field(default_factory=list, description="Enabled tool IDs")
    knowledge_base_id: Optional[str] = Field(default=None, description="Knowledge base ID")
    memory_enabled: bool = Field(default=False, description="Enable long-term memory")

    # Team role
    domain: Optional[str] = Field(default=None, max_length=100, description="Expertise domain")
    collaboration_style: str = Field(
        default="supportive",
        pattern="^(dominant|supportive|critical)$",
        description="Collaboration style",
    )
    speaking_priority: int = Field(default=5, ge=1, le=10, description="Speaking priority (1-10)")
    interaction_rules: InteractionRules = Field(
        default_factory=InteractionRules,
        description="Interaction rules",
    )


class AgentCreate(AgentBase):
    """Schema for creating a new agent."""

    is_template: bool = Field(default=False, description="Whether this is a template")
    is_public: bool = Field(default=False, description="Whether publicly visible")


class AgentUpdate(BaseModel):
    """Schema for updating an agent. All fields are optional."""

    name: Optional[str] = Field(default=None, min_length=1, max_length=100)
    avatar: Optional[str] = Field(default=None, max_length=500)
    description: Optional[str] = None
    tags: Optional[list[str]] = None

    system_prompt: Optional[str] = Field(default=None, min_length=1)
    model_id: Optional[str] = None
    temperature: Optional[float] = Field(default=None, ge=0, le=2)
    max_tokens: Optional[int] = Field(default=None, ge=1, le=128000)

    tools: Optional[list[str]] = None
    knowledge_base_id: Optional[str] = None
    memory_enabled: Optional[bool] = None

    domain: Optional[str] = Field(default=None, max_length=100)
    collaboration_style: Optional[str] = Field(default=None, pattern="^(dominant|supportive|critical)$")
    speaking_priority: Optional[int] = Field(default=None, ge=1, le=10)
    interaction_rules: Optional[InteractionRules] = None

    is_public: Optional[bool] = None


class AgentResponse(AgentBase):
    """Schema for agent response."""

    id: str
    user_id: str
    version: int
    is_template: bool
    is_public: bool
    parent_id: Optional[str]

    # Statistics
    usage_count: int
    rating: float
    rating_count: int

    # Timestamps
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True


class AgentListResponse(BaseModel):
    """Schema for agent list response (simplified)."""

    id: str
    name: str
    avatar: Optional[str]
    description: Optional[str]
    tags: list[str]
    domain: Optional[str]
    collaboration_style: str
    is_template: bool
    is_public: bool
    usage_count: int
    rating: float
    created_at: datetime

    class Config:
        from_attributes = True


class AgentFromTemplate(BaseModel):
    """Schema for creating agent from template."""

    template_id: str = Field(..., description="Template agent ID to copy from")
    name: str = Field(..., min_length=1, max_length=100, description="New agent name")
    customizations: Optional[AgentUpdate] = Field(
        default=None,
        description="Optional customizations to apply",
    )

File: backend/src/models/agent.rs
use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct InteractionRules {
    #[serde(default = "default_true")]
    pub can_challenge: bool,
    #[serde(default = "default_true")]
    pub can_be_challenged: bool,
    #[serde(default)]
    pub defer_to: Vec<String>,
}

impl Default for InteractionRules {
    fn default() -> Self {
        Self {
            can_challenge: true,
            can_be_challenged: true,
            defer_to: Vec::new(),
        }
    }
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Agent {
    pub id: String,
    pub user_id: String,
    pub name: String,
    pub avatar: Option<String>,
    pub description: Option<String>,
    #[serde(default)]
    pub tags: Vec<String>,
    pub system_prompt: String,
    pub model_id: Option<String>,
    pub temperature: f64,
    pub max_tokens: u32,
    #[serde(default)]
    pub tools: Vec<String>,
    pub knowledge_base_id: Option<String>,
    pub memory_enabled: bool,
    pub domain: Option<String>,
    pub collaboration_style: String,
    pub speaking_priority: i32,
    #[serde(default)]
    pub interaction_rules: InteractionRules,
    pub version: u32,
    pub is_template: bool,
    pub is_public: bool,
    pub parent_id: Option<String>,
    pub usage_count: u32,
    pub rating: f64,
    pub rating_count: u32,
    pub created_at: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct AgentListItem {
    pub id: String,
    pub name: String,
    pub avatar: Option<String>,
    pub description: Option<String>,
    #[serde(default)]
    pub tags: Vec<String>,
    pub domain: Option<String>,
    pub collaboration_style: String,
    pub is_template: bool,
    pub is_public: bool,
    pub usage_count: u32,
    pub rating: f64,
    pub created_at: DateTime<Utc>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct AgentCreate {
    pub name: String,
    #[serde(default)]
    pub avatar: Option<String>,
    #[serde(default)]
    pub description: Option<String>,
    #[serde(default)]
    pub tags: Vec<String>,
    pub system_prompt: String,
    #[serde(default)]
    pub model_id: Option<String>,
    #[serde(default = "default_temperature")]
    pub temperature: f64,
    #[serde(default = "default_max_tokens")]
    pub max_tokens: u32,
    #[serde(default)]
    pub tools: Vec<String>,
    #[serde(default)]
    pub knowledge_base_id: Option<String>,
    #[serde(default)]
    pub memory_enabled: bool,
    #[serde(default)]
    pub domain: Option<String>,
    #[serde(default = "default_collaboration_style")]
    pub collaboration_style: String,
    #[serde(default = "default_speaking_priority")]
    pub speaking_priority: i32,
    #[serde(default)]
    pub interaction_rules: InteractionRules,
    #[serde(default)]
    pub is_template: bool,
    #[serde(default)]
    pub is_public: bool,
}

#[derive(Debug, Clone, Serialize, Deserialize, Default)]
pub struct AgentUpdate {
    pub name: Option<String>,
    pub avatar: Option<String>,
    pub description: Option<String>,
    pub tags: Option<Vec<String>>,
    pub system_prompt: Option<String>,
    pub model_id: Option<String>,
    pub temperature: Option<f64>,
    pub max_tokens: Option<u32>,
    pub tools: Option<Vec<String>>,
    pub knowledge_base_id: Option<String>,
    pub memory_enabled: Option<bool>,
    pub domain: Option<String>,
    pub collaboration_style: Option<String>,
    pub speaking_priority: Option<i32>,
    pub interaction_rules: Option<InteractionRules>,
    pub is_public: Option<bool>,
}

fn default_true() -> bool {
    true
}

fn default_temperature() -> f64 {
    0.7
}

fn default_max_tokens() -> u32 {
    2048
}

fn default_speaking_priority() -> i32 {
    5
}

fn default_collaboration_style() -> String {
    "supportive".to_string()
}

